File: Adivine_la_Estructura.cpp
#include <iostream>
#include <queue>
#include <stack>
using namespace std;
 
int main(int argc, char *argv[])
{
    int n, c, d, ele;
    bool sw1, sw2, sw3;
    while (cin >> n)
    {
        sw1 = true;
        sw2 = true;
        sw3 = true;
        stack <int> p;
        queue<int> q;
        priority_queue<int> pq;
        while (n--)
        {
            cin >> c >> d;
            if (c == 1)
            {
                p.push(d);
                q.push(d);
                pq.push(d);
            }
            else
            {
                if (!p.empty() && sw1)
                {
                    ele = p.top();
                    p.pop();
                    if (d != ele)
                        sw1 = false;
                }
                else
                    sw1 = false;
                if (!q.empty() && sw2)
                {
                    ele = q.front();
                    q.pop();
                    if (d != ele)
                        sw2 = false;
                }
                else
                    sw2 = false;
                if (!pq.empty() && sw3)
                {
                    ele = pq.top();
                    pq.pop();
                    if (d != ele)
                        sw3 = false;
                }
                else
                    sw3 = false;
            }
        }
        if (sw1 && !sw2 && !sw3)
            cout << "stack";
        else if (!sw1 && sw2 && !sw3)
            cout << "queue";
        else if (!sw1 && !sw2 && sw3)
            cout << "priority queue";
        else if (!sw1 && !sw2 && !sw3)
            cout << "impossible";
        else
            cout << "not sure";
        cout << '\n';
    }
}

File: Adivine_la_Estructura.py
#1046
from collections import deque
from queue import PriorityQueue
import sys
def main():
    n = 1; c = 0; d=0; ele = 0
    sw1 = False; sw2 = False; sw3 = False
    for linea in sys.stdin:
        if str(linea) == "\n":
            break
        else:
            sw1 = True
            sw2 = True
            sw3 = True
            p = []
            q = deque([])
            pq = PriorityQueue()
            n = int(linea)
            while n > 0:
                c, d = map(int, input().split())
                if c == 1:
                    p.append(d)
                    q.append(d)
                    pq.put(d)
                else:
                    if len(p) >0 and sw1:
                        ele = p[-1]
                        p.pop()
                        if d != ele:
                            sw1 = False
                    else:
                        sw1 = False
                        
                    if len(q)>0 and sw2:
                        ele = q[0]
                        q.popleft()
                        if d != ele:
                            sw2 = False
                    else:
                        sw2 = False
                    if not pq.empty() and sw3:
                        ele = pq.queue[-1]
                        pq.queue.pop()
                        if d != ele:
                            sw3 = False
                    else:
                        sw3 = False
                n -= 1
            if sw1 and not sw2 and not sw3:
                print("stack")
            elif not sw1 and sw2 and not sw3:
                print("queue")
            elif not sw1 and not sw2 and sw3:
                print("priority queue")
            elif not sw1 and not sw2 and not sw3:
                print("impossible")
            else:
                print("not sure")
main()
                
            
            